chore: provide a better out-of-gas error UX

D4ryl00 opened the pull request:
fixes https://github.com/gnolang/gno/issues/3704

This PR reworks gas simulation and out-of-gas UX end-to-end, with a clear split of responsibilities:

- **Server**: executes tx and reports consistent out-of-gas (OOG) errors.
- **Client (`gnokey`)**: chooses simulation gas ceiling and provides guidance to users.

## High-level changes

- Moved simulation gas-ceiling behavior to the **client**.
  - `gnokey` queries consensus max gas and uses it for simulation.
  - If unavailable, it uses a large fallback ceiling.
- Standardized OOG messaging across execution paths.
  - Errors now explain whether the limit hit was tx gas wanted vs max block gas.
  - Messages include operation context and simulation guidance when relevant.
- Improved simulation feedback.
  - Adds a `suggested gas-wanted` value (+5% margin).
  - Preserves fee estimation info in output.
- Improved failed-tx output UX in CLI.
  - Failure path now prints the same useful metrics users see on success.

## User impact

- Better gas estimation workflow.
- Clearer and more actionable OOG errors.
- More predictable behavior between `simulate only` and `simulate skip` (with explicit messaging when they differ due to simulation ceiling behavior).

## Usage

### Before

```
gnokey maketx call -pkgpath "gno.land/r/demo/counter" -func "Increment" -broadcast -gas-wanted=105000 -gas-fee=500ugnot test1
Enter password.
TX HASH:
INFO:
--= Error =--
Data: out of gas error
Msg Traces:
    0  gno/tm2/pkg/errors/errors.go:103 - deliver transaction failed: log:out of gas, gasWanted: 105000, gasUsed: 110998 location: GetObjectPerByte
--= /Error =--
```

GAS USED is the gas used until the panic during the simulation stage.

### Now

```
gnokey maketx call -pkgpath "gno.land/r/demo/counter" -func "Increment" -broadcast -gas-wanted=105000 -gas-fee=500ugnot test1
Enter password.
GAS WANTED: 105000
GAS USED:   1732814
HEIGHT:     0
STORAGE DELTA:  5 bytes
STORAGE FEE:    500ugnot
TOTAL TX COST:  1000ugnot
EVENTS:     [{"bytes_delta":5,"fee_delta":{"denom":"ugnot","amount":500},"pkg_path":"gno.land/r/demo/counter"}]
INFO:       suggested gas-wanted (gas used + 5%): 1819455
TX HASH:
--= Error =--
Data: out of gas error
Msg Traces:
    0  github.com/gnolang/gno/tm2/pkg/errors/errors.go:103 - deliver transaction failed: log:gas used (1732814) exceeds tx's gas wanted (105000) during operation: simulation
--= /Error =--
```

GAS USED is the total gas used of the transaction, now calculate during the simulation stage.

```
gnokey maketx call -pkgpath "gno.land/r/demo/counter" -func "Increment" -broadcast -gas-wanted=105000 -gas-fee=500ugnot -simulate skip test1
Enter password.
--= Error =--
Data: out of gas error
Msg Traces:
    0  github.com/gnolang/gno/tm2/pkg/errors/errors.go:103 - check transaction failed: log:--= Error =--
Data: std.OutOfGasError{abciError:std.abciError{}}
Msg Traces:
    0  gno/tm2/pkg/std/errors.go:117 - gas used (179810) exceeds tx's gas wanted (105000) during operation: DepthReadFlat; simulate with consensus maximum (10000000000) to get real transaction usage
Stack Trace:
    0  gno/tm2/pkg/errors/errors.go:93
...
```

Message updated: GAS USED is the gas used UNTIL the panic (no simulation).

```
gnokey maketx call -pkgpath "gno.land/r/demo/counter" -func "Increment" -broadcast -gas-wanted=100 -gas-fee=500ugnot -simulate skip test1
Enter password.
--= Error =--
Data: out of gas error
Msg Traces:
    0  github.com/gnolang/gno/tm2/pkg/errors/errors.go:103 - check transaction failed: log:--= Error =--
Data: std.OutOfGasError{abciError:std.abciError{}}
Msg Traces:
    0  gno/tm2/pkg/std/errors.go:117 - gas used (2800) exceeds tx's gas wanted (100) during operation: txSize; simulate with consensus maximum (10000000000) to get real transaction usage
Stack Trace:
    0  gno/tm2/pkg/errors/errors.go:93
    ...
```

Message updated: GAS USED is the gas used UNTIL the panic (before running the tx).

D4ryl00 committed:
chore: provide a better out-of-gas error UX

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

Gno2D2 commented:
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)
- [x] The pull request description provides enough details (checked by @thehowl)


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🟢 Pending initial approval by a review team member, or review from tech-staff


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://www.github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 The pull request was created from a fork (head branch repo: D4ryl00/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>Pending initial approval by a review team member, or review from tech-staff</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 Not (🔴 Pull request author is a member of the team: tech-staff)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 If
    ├── 🟢 Condition
    │   └── 🟢 Or
    │       ├── 🟢 User jefft0 already reviewed PR 4951 with state APPROVED
    │       ├── 🟢 At least 1 user(s) of the team tech-staff reviewed pull request
    │       └── 🔴 This pull request is a draft
    └── 🟢 Then
        └── 🟢 Not (🔴 This label is applied to pull request: review/triage-pending)

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

<details><summary>The pull request description provides enough details</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 Not (🔴 Pull request author is a member of the team: core-contributors)
    └── 🟢 Not (🔴 Pull request author is user: dependabot[bot])

```
### Can be checked by
 - team core-contributors

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

D4ryl00 committed:
chore: update error message for tx execution error

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 committed:
fix: use the consonsus MaxGas to setup the gas meter in tx simulation

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 committed:
chore: add unit tests

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 committed:
chore: add more tests

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 committed:
chore: add more tests

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

jefft0 reviewed (approved):
The PR resolves the related issue. The solution is documented in the PR description. All tests pass. Many new tests to resolve codecov. Ready for core dev review.

davd-gzl reviewed (approved):
I've tested every situations, and it all works correctly to me.
Implementation LGTM

D4ryl00 committed:
chore: improve message error

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 committed:
feat: add suggested gas-wanted

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 committed:
chore: in gas-wanted suggestion message, replace ; by ,

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 commented on tm2/pkg/sdk/baseapp.go:
```diff
@@ -762,16 +762,26 @@ func (app *BaseApp) runTx(ctx Context, tx Tx) (result Result) {
+				}
 				log := fmt.Sprintf(
-					"out of gas, gasWanted: %d, gasUsed: %d location: %v",
+					"out of gas %s, gasWanted: %d, gasUsed: %d (until panic) location: %v",
```
The log isn’t trying to guess a new gas value; it just reports where execution stopped. To get a reliable number to set `gas-wanted`, you should re-run in `--simulate only`, which reports the full gasUsed when it completes. This work was aimed at clarifying that the error’s gasUsed is “until panic,” not a recommendation. If a user skips simulation, they’re accepting that the logged gasUsed may be partial.

D4ryl00 committed:
Update tm2/pkg/sdk/auth/ante.go

Co-authored-by: Morgan <morgan@morganbaz.com>

D4ryl00 committed:
fix: txtar test

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

thehowl commented on tm2/pkg/sdk/auth/ante.go:
```diff
@@ -408,6 +408,22 @@ func EnsureSufficientMempoolFees(ctx sdk.Context, fee std.Fee) sdk.Result {
 
 // SetGasMeter returns a new context with a gas meter set from a given context.
 func SetGasMeter(ctx sdk.Context, gasLimit int64) sdk.Context {
+	if ctx.Mode() == sdk.RunTxModeSimulate {
```
You mean, as a request on what to do?

I think that swapping the max-gas to the consensus' maximum should be done on the client, not on the server. The server simulation should match as much as possible what would happen; if the client wants to calculate the gas required, then it's the client's responsibility to set the gas wanted to the consensus maximum.

thehowl commented on tm2/pkg/sdk/baseapp.go:
```diff
@@ -762,16 +763,26 @@ func (app *BaseApp) runTx(ctx Context, tx Tx) (result Result) {
 		if r := recover(); r != nil {
 			switch ex := r.(type) {
 			case store.OutOfGasError:
```
I would go for a different message:

```
gas used (%d) exceeds tx's gas wanted (%d) during operation: %v; simulate with consensus maximum (%d) to get real transaction usage
```

when gas-wanted == consensus maximum, substitute "tx's gas wanted" with "max block gas" and remove the last sentence.

D4ryl00 replied:
Implemented in 0fc2453e3ad03e7393424fbcb8ac6ec934228b27
New behavior:

- If maxGas > 0 and gasWanted != maxGas:
    - gas used (%d) exceeds tx's gas wanted (%d) during operation:
      %v; simulate with consensus maximum (%d) to get real transaction usage
- If maxGas > 0 and gasWanted == maxGas:
    - gas used (%d) exceeds max block gas (%d) during operation: %v
    - no simulation sentence
- If consensus max gas is unavailable/non-positive, it falls back to:
    - gas used (%d) exceeds tx's gas wanted (%d) during operation: %v

thehowl reviewed (commented):
Here's my take:

- We should remove the changes in sdk/auth; as well as those in baseapp not relating to the message. Reason: we should not change how simulations work. A simulation is an attempt to run a transaction as close as possible to how it would run on-chain; so the environment shouldn't change.
- We can change the client package to make it so that when doing a simulation, it swaps by default the gas-wanted with the transaction gas-wanted. This may require an additional RPC call (to get the consensus max-gas); if possible, let's run it in a separate goroutine at the same time we get ie. the account number from the chain, so there's no signficant slowdown on gnokey.

D4ryl00 committed:
feat: the client changes the max gas, not the server

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 committed:
feat: improve OOG error logs in ante and baseapp

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 commented:
> Here's my take:
> 
>     * We should remove the changes in sdk/auth; as well as those in baseapp not relating to the message. Reason: we should not change how simulations work. A simulation is an attempt to run a transaction as close as possible to how it would run on-chain; so the environment shouldn't change.
> 
>     * We can change the client package to make it so that when doing a simulation, it swaps by default the gas-wanted with the transaction gas-wanted. This may require an additional RPC call (to get the consensus max-gas); if possible, let's run it in a separate goroutine at the same time we get ie. the account number from the chain, so there's no signficant slowdown on gnokey.

Implemented in 9492802370985dd3a64e773ef2d59346d1b76476. I reverted baseapp and ante so they don't change how the simulation works. It's now the client that handles it.

D4ryl00 committed:
fix: txtar gas value

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 committed:
fix: txtar tests

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 commented:
@thehowl 
The downside of that "client" implementation is that the GasUsed differ because simulate only and simulate skip are running on different tx bytes now.

  - In simulate only, the client rewrites GasWanted to consensus max before simulation.
  - In simulate skip, it broadcasts the original tx bytes directly .

  Gas includes tx-size cost in ante:

  - tm2/pkg/sdk/auth/ante.go charges TxSizeCostPerByte * len(txBytes).

  So if rewritten GasWanted changes encoded tx size by 1 byte, you get a +10 gas delta.
  This is expected with the current client-side “simulate with consensus max gas” behavior.

D4ryl00 committed:
chore: refactor OutOfGasLog, margin overflow and buildSimulationTxBytes

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 committed:
Merge remote-tracking branch 'upstream/master' into fix/out-of-gas-error-amount

D4ryl00 committed:
fix: gas amount

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 committed:
Merge remote-tracking branch 'upstream/master' into fix/out-of-gas-error-amount

thehowl commented on tm2/pkg/crypto/keys/client/broadcast.go:
```diff
@@ -124,7 +129,24 @@ func BroadcastHandler(cfg *BroadcastCfg) (*ctypes.ResultBroadcastTxCommit, error
+			}
+		}
+		if res != nil {
+			appendSuggestedGasWanted(res)
```
It's not about just showing it there, but it's about the natural logic. Currently an output would say:

```
suggested gas-wanted (gas used + 5%): 2955207, estimated gas usage: 2814482, ...
```

This feels unnatural; it's better to say something like:

```
estimated gas usage: 2814482 (suggested, with 5% margin: 2955207), gas fee: ...
```

D4ryl00 replied:
[3fbfe9f](https://github.com/gnolang/gno/pull/4951/commits/3fbfe9f7acc3ad233d2b957508bab1654985ff8c)

thehowl commented on tm2/pkg/crypto/keys/client/maketx.go:
```diff
@@ -241,3 +259,44 @@ func ExecSignAndBroadcast(
+		return 0, err
+	}
+
+	res, err := cli.ConsensusParams(context.TODO(), nil)
```
good to make context.background; maybe note that httpclient has implicitly a 60s-timeout by default

D4ryl00 replied:
[aceed43](https://github.com/gnolang/gno/pull/4951/commits/aceed4391654548066da170a8cb639644be16a83)

thehowl commented on tm2/pkg/crypto/keys/client/maketx.go:
```diff
@@ -174,13 +185,22 @@ func SignAndBroadcastHandler(
+	var maxGas int64
+	if maxGasCh != nil {
+		res := <-maxGasCh
+		if res.err == nil {
```
so, I actually don't think we should swallow res.err: we will query the same RPC in just one second with a likely more complex result. we should at least warn that we attempted to get the max gas and failed

D4ryl00 replied:
[8a41756](https://github.com/gnolang/gno/pull/4951/commits/8a4175673b952f0edc5faf78d4a216ea06d55df4)

thehowl commented on tm2/pkg/crypto/keys/client/broadcast.go:
```diff
@@ -143,6 +166,43 @@ func BroadcastHandler(cfg *BroadcastCfg) (*ctypes.ResultBroadcastTxCommit, error
+func buildSimulationTxBytes(tx *std.Tx, txBytes []byte, maxGas int64) ([]byte, bool, error) {
+	if maxGas <= 0 {
+		maxGas = simulationMaxGasFallback
+	}
```
also, I think this should be removed:

maxGas is likely to be `<= 0` only when we fail fetching the max block gas. thus setting it to `MaxInt64` would automatically lead to an error if we were to try to simulate the transaction; namely:

https://github.com/gnolang/gno/blob/066f15a79e0c0889ff21b2899f666a926994ebd1/tm2/pkg/sdk/auth/ante.go#L54

on the other hand, the transaction _can_ succeed if we just use `tx.Fee.GasWanted`; so we can rely on the condition afterwards to simply avoid trying to modify the transaction's gas wanted.

D4ryl00 replied:
[1090071](https://github.com/gnolang/gno/pull/4951/commits/1090071270f95ec4ee78288ceb8bce53c573c402)
I didn't remove it at all because it could make sense to test if maxGas == -1 to put simulationMaxGasFallback since in ante.go it doesn't check when maxGas == -1.
Let me know what do you think.

thehowl reviewed (approved):
a few nits; then I would say we're good for a review by @gfanton or @alexiscolin

jefft0 committed:
Merge branch 'master' into fix/out-of-gas-error-amount

jefft0 committed:
chore: gnokey_gasfee.txtar: Use new gas amounts

Signed-off-by: Jeff Thompson <jeff@thefirst.org>

D4ryl00 committed:
chore: refactor OOG message logic

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 committed:
chore: change context.TODO() by context.Background()

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 committed:
chore: warn if cannot fetch consensus max gas

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 committed:
chore: rewrite buildSimulationTxBytes maxGas check

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 committed:
Merge remote-tracking branch 'upstream/master' into fix/out-of-gas-error-amount

D4ryl00 committed:
fix: CI lint

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 committed:
fix: infinite_loop.txtar timeout fix

Signed-off-by: D4ryl00 <d4ryl00@gmail.com>

D4ryl00 committed:
Merge remote-tracking branch 'upstream/master' into fix/out-of-gas-error-amount

thehowl committed:
Merge branch 'master' into fix/out-of-gas-error-amount
